### ISSUE TITLE
Cannot implicitly convert type 'System.Threading.Tasks.Task<SomeType>' to 'System.Threading.Tasks.Task<Moq.It.IsAnyType>'

### ISSUE THREAD

**Issue by bhuvesh321** [devlooped/moq#1124]
I am using the following setup to mock a generic method

```csharp
Task<T> GetDataAsync<T>(string arg);
mock
.Setup(t => t.GetData<It.IsAnyType>(It.IsAny<string>())
.Returns((string arg)=>{
  // some logic
  return Task<actualType>; // This gives error cannot convert to type It.IsAnyType
});
```

How to solve this?

**Comment by stakx** [comment]
TL;DR: See the first code block below.

First, let's look at why you need `It.IsAnyType` at all.

In theory, any type in .NET is convertible to `object`, so if you set up `GetDataAsync<object>` instead of `GetDataAsync<It.IsAnyType>`, your setup will still trigger for any type argument.

However, that is wrong because the static typing then suggests that it is sufficient for your `.Returns` callback to return a `Task<object>` instance... which won't actually work, depending on how your method gets invoked: If I'm calling `mock.Object.GetDataAsync<bool>`, I expect to get back a `Task<bool>`, not a `Task<object>`.

So you are right to set up `GetDataAsync<It.IsAnyType>`.

However, now your `.Returns` callback has the somewhat tricky task of having to produce a `Task<>` instance of an unknown type.

If you continue using `.Returns((string arg) => ...)`, you have no way of knowing what concrete type the caller has used in place of `It.IsAnyType`. Also, static typing dictates that your callback should return an instance of `Task<It.IsAnyType>`, which is impossible.

Therefore, you need to use a different `.Returns` overload that allows you to get at the concrete type argument, and return an untyped `object`. That overload is `.Returns(InvocationFunc)`.

With `.Returns(InvocationFunc)`, your callback receives an `IInvocation` object that represents the current call. From that, you can access two properties that are of interest here: `Method` and `Arguments`.

* `Method` is of type `MethodInfo`, so you can get at the generic type arguments here using `.GetGenericArguments()`.
* `Arguments` is an `object[]` array containing the regular arguments (e.g. the `arg` string).

So you'll likely end up with something similar to the following:

```csharp
mock.Setup(m => m.GetDataAsync<It.IsAnyType>(It.IsAny<string>()))
    .Returns(new InvocationFunc(invocation =>
    {
        // the first argument is the value that the caller provided for the `string arg` parameter:
        var arg = (string)invocation.Arguments[0];

        // the first generic type argument is whatever the caller chose in place of `It.IsAnyType`:
        var resultType = invocation.Method.GetGenericArguments()[0];

        // now you need to produce a Task<resultType>, for example:
        var defaultValue = resultType.IsValueType ? Activator.CreateInstance(resultType) : null;
        var taskFromResultMethod = typeof(Task).GetMethod("FromResult", BindingFlags.Public | BindingFlags.Static).MakeGenericMethod(resultType);
        return taskFromResultMethod.Invoke(null, new[] { defaultValue });
    }));
```

(While it may be a little annoying to lose static typing for the regular arguments `(string arg)`, you would at some point have lost static typing anyway due to the statically unknown return type.)

Let's see if that works:

```csharp
var task = mock.Object.GetDataAsync<bool>("Grog");
Console.WriteLine(task.GetType());  // => System.Threading.Tasks.Task`1[System.Boolean]

var result = await task;
Console.WriteLine(result);  // => False
```

Looks good.

(But if all you want to do is produce a completed `Task<>` with a default value as the result, you could've just left all that work to the `DefaultValue` provider of your mock and put some other logic in a `.Callback` instead.)

I hope this answers your question.

Have a happy new year! 🎆
